selectolax/utils.pxi

```diff
@@ -11,3 +17,91 @@ def preprocess_input(html, decode_errors='ignore'):
     if html_len > MAX_HTML_INPUT_SIZE:
         raise ValueError("The specified HTML input is too large to be processed (%d bytes)" % html_len)
     return bytes_html, html_len
+
+
+def do_create_tag(tag: str, parser_cls: ParserCls):
+    if not tag:
+        raise ValueError("Tag name cannot be empty")
+    return do_parse_fragment(f"<{tag}></{tag}>", parser_cls)[0]
+
+
+def get_fragment_type(
+    html: str,
+    parser_cls: ParserCls,
+    tree: Optional[Parser] = None,
+) -> Literal["document", "fragment", "head", "body", "head_and_body", "document_no_head", "document_no_body", "document_no_head_no_body"]:
```

Comment:
Because of how lexbor / modest adds the `<html>` and other tags, then each case requires a bit different handling for when `<html>` / `<head>` / `<body>` tags ARE or ARE NOT present. So I've split it in 2 - first a function that decides the state, and the second that post-processes the parsed HTML tree given the state.